javascript.md

```diff
@@ -0,0 +1,793 @@
+      console.log(values[i]);
+    };
+  }(i), 1000 * i);
+}
+```
+
+##### Good
+
+```js
+var values = [1, 2, 3];
+var i;
+
+for (i = 0; i < values.length; i++) {
+  setTimeout(function (i) {
+    console.log(values[i]);
+  }, 1000 * i, i);
+}
+```
+
+```js
+var values = [1, 2, 3];
+var i;
+
+for (i = 0; i < values.length; i++) {
+  wait(i);
+}
+
+function wait (i) {
+  setTimeout(function () {
+    console.log(values[i]);
+  }, 1000 * i);
+}
+```
+
+Or even better, just use `.forEach` which doesn't have the same caveats as declaring functions in `for` loops.
+
+##### Better
+
+```js
+[1, 2, 3].forEach(function (value, i) {
+  setTimeout(function () {
+    console.log(value);
+  }, 1000 * i);
+});
+```
+
+Whenever a method is non-trivial, make the effort to **use a named function declaration rather than an anonymous function**. This will make it easier to pinpoint the root cause of an exception when analyzing stack traces.
+
+##### Bad
+
+```js
+function once (fn) {
+  var ran = false;
+  return function () {
+    if (ran) { return };
+    ran = true;
+    fn.apply(this, arguments);
+  };
+}
+```
+
+##### Good
+
+```js
+function once (fn) {
+  var ran = false;
+  return function run () {
+    if (ran) { return };
+    ran = true;
+    fn.apply(this, arguments);
+  };
+}
+```
+
+Avoid keeping indentation levels from raising more than necessary by using guard clauses instead of flowing `if` statements.
+
+##### Bad
+
+```js
+if (car) {
+  if (black) {
+    if (turbine) {
+      return 'batman!';
+    }
+  }
+}
+```
+
+```js
+if (condition) {
+  // 10+ lines of code
+}
+```
+
+##### Good
+
+```js
+if (!car) {
+  return;
+}
+if (!black) {
+  return;
+}
+if (!turbine) {
+  return;
+}
+return 'batman!';
+```
+
+```js
+if (!condition) {
+  return;
+}
+// 10+ lines of code
+```
+
+## Prototypes
+
+Hacking native prototypes should be avoided at all costs, use a method instead. If you must extend the functionality in a native type, try using something like [poser][23] instead.
+
+##### Bad
+
+```js
+String.prototype.half = function () {
+  return this.substr(0, this.length / 2);
+};
+```
+
+##### Good
+
+```js
+function half (text) {
+  return text.substr(0, text.length / 2);
+}
+```
+
+**Avoid prototypical inheritance models** unless you have a very good _performance reason_ to justify yourself.
+
+- Prototypical inheritance boosts puts need for `this` through the roof
+- It's way more verbose than using plain objects
+- It causes headaches when creating `new` objects
+- Needs a closure to hide valuable private state of instances
+- Just use plain objects instead
+
+## Object Literals
+
+Instantiate using the egyptian notation `{}`. Use factories instead of constructors, here's a proposed pattern for you to implement objects in general.
+
+```js
+function util (options) {
+  // private methods and state go here
+  var foo;
+
+  function add () {
+    return foo++;
+  }
+
+  function reset () { // note that this method isn't publicly exposed
```

Comment:
Personally I'd underscore prefix this method to show that it's not being exposed externally:

```js
function _reset() {
...
```